### ISSUE TITLE
println!  have different behave for similar data 

### ISSUE THREAD

**Issue by ser-vasilich** [rust-lang/rust#96522]
fn main() {
    let a=-1.14;
    let b=a+1.0;
    println!("{}",a);
    println!("{}",b);
    
    println!("{:.18}",a);
    println!("{:.18}",b);
}

Standard Output

-1.14
-0.1399999999999999
-1.139999999999999902
-0.139999999999999902

**Comment by NinoScript** [comment]
That looks fine to me, [floats have finite precision](https://docs.microsoft.com/en-us/cpp/build/why-floating-point-numbers-may-lose-precision), and that happens in all programming languages I've used.

Are you expecting something different?

**Comment by wwylele** [comment]
Just to explain this better, let's add a third print of the "expected result" `let c=-0.14`
```
fn main() {
  let a=-1.14;
  let b=a+1.0;
  let c=-0.14;
  println!("{}",a);
  println!("{}",b);
  println!("{}",c);
  
  println!("{:.18}",a);
  println!("{:.18}",b);
  println!("{:.18}",c);

}
```
```
-1.14
-0.1399999999999999
-0.14
-1.139999999999999902
-0.139999999999999902
-0.140000000000000013
```
As you can see, `println!` prints `-0.14` fine. The problem here is that in the float universe, `-0.140000000000000013` is a closer to `-0.14` than `-0.139999999999999902` is, so the former is chosen as the canonical representation of `-0.14`. `let b = a+1.0;` Unfortunately doesn't land on that canonical representation.

In short, this is not a problem in `println!`. Rather, it is because of how float is stored and how `+` is implemented.

**Comment by ser-vasilich** [comment]
For me not fine. 
See the next example. Only default format out not similar result.

fn main() {
    let a=-1.14;
    let b=a+1.0;
    println!("Default not same:");
    println!("{}",a);
    println!("{}",b);
    
    for i in 1..20 {
      println!("Same:");
      println!("{:.1$}",a,i);
      println!("{:.1$}",b,i);
    }
}

Default not same:
-1.14
-0.1399999999999999
Same:
-1.1
-0.1
Same:
-1.14
-0.14
Same:
-1.140
-0.140
Same:
-1.1400
-0.1400
Same:
-1.14000
-0.14000
Same:
-1.140000
-0.140000
Same:
-1.1400000
-0.1400000
Same:
-1.14000000
-0.14000000
Same:
-1.140000000
-0.140000000
Same:
-1.1400000000
-0.1400000000
Same:
-1.14000000000
-0.14000000000
Same:
-1.140000000000
-0.140000000000
Same:
-1.1400000000000
-0.1400000000000
Same:
-1.14000000000000
-0.14000000000000
Same:
-1.140000000000000
-0.140000000000000
Same:
-1.1399999999999999
-0.1399999999999999
Same:
-1.13999999999999990
-0.13999999999999990
Same:
-1.139999999999999902
-0.139999999999999902
Same:
-1.1399999999999999023
-0.1399999999999999023

**Comment by ser-vasilich** [comment]
I expected similar behavior: different in 1 digit.
I know programming language which out adequate result:

**Comment by ser-vasilich** [comment]
Why does only default formatting has not similar results?

**Comment by wwylele** [comment]
The default printing is "print enough precision such that when you copy it back as a code (i.e. do a float -> string -> float roundtrip), it still ends up the exact float"

If we cut the default precision of `-1.14 + 1.0` short to just `"-0.14"`, the round trip will be

`-0.139999999999999902` -> `"-1.14"` -> `-0.140000000000000013`

You end up with a different number through the roundtrip 

`-1.14` itself doesn't have this problem, because there is no float number `-1.140000000000000013`, and such round trip still preserve the value with short precision.

**Comment by ser-vasilich** [comment]
You didn't convince me. I see the difference between addition and subtraction, or selective rounding. If you do not see - consider the example:

**Comment by wwylele** [comment]
The last example is just the classical `c + d - d != c` in floating point numbers. (and `println!` faithfully prints different things for both sides) There are many online resource online teaching this behavior.
